### PR TITLE
New module: Add module to configure inbound email reception with Amazon Simple Email (cloud/amazon/ses_rule)

### PR DESCRIPTION
**Pull request by tomislacker**
##### SUMMARY
Add a pair of modules to support [configuring inbound email reception](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email.html) with Amazon's Simple Email Service.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME

ses_rule
ses_rule_set

##### ANSIBLE VERSION

```
ansible 2.2.0.0 (local cdec853e37) last updated 2016/11/15 14:57:04 (GMT -500)
  lib/ansible/modules/core: (detached HEAD fe9c56a003) last updated 2016/11/16 15:05:06 (GMT -500)
  lib/ansible/modules/extras: (detached HEAD f564e77a08) last updated 2016/11/16 15:05:07 (GMT -500)
  config file = 
  configured module search path = Default w/o overrides
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ses_rule.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    check_mode = module.check_mode
+    changed = False
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='Python module "boto3" is missing, please install it')
+
+    if not HAS_BOTOCORE:
+        module.fail_json(msg='Python module "botocore" is missing, please install it')
+
+    if len(recipients) == 0 and state == 'present':
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ses_rule.py]: Use `required_if(['state', 'present', ['recipients', 'actions']])` in the `AnsibleModule` arguments

File: lib/ansible/modules/cloud/amazon/ses_rule.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+RETURN = ""
+
+try:
+    import boto3
+    HAS_BOTO3 = True
+except ImportError:
+    HAS_BOTO3 = False
+
+try:
+    import botocore
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ses_rule.py]: You can import HAS_BOTO3, and you don't need to import boto3. You can also assume that boto3 and botocore are installed at the same time, and then it just looks something like this:

```
from ansible.module_utils.ec2 import HAS_BOTO3

try:
    import botocore
except ImportError
    pass  # caught by HAS_BOTO3
```

File: lib/ansible/modules/cloud/amazon/ses_rule_set.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    rule_sets = ses_client.list_receipt_rule_sets()['RuleSets']
+    return len([s for s in rule_sets if s['Name'].lower() == name.lower()]) > 0
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        name=dict(type='str', required=True),
+        state=dict(type='str', default='present', choices=['present', 'absent']),
+        activate=dict(type='bool', default=False),
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ses_rule_set.py]: Should be an adjective not a verb  -  `active` in this case seems to meet how it's used later on.

File: lib/ansible/modules/cloud/amazon/ses_rule_set.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+try:
+    import botocore
+    HAS_BOTOCORE = True
+except ImportError:
+    HAS_BOTOCORE = False
+
+
+def rule_set_exists(ses_client, name):
+    rule_sets = ses_client.list_receipt_rule_sets()['RuleSets']
+    return len([s for s in rule_sets if s['Name'].lower() == name.lower()]) > 0
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ses_rule_set.py]: `any(X)` is probably preferable to `len(X) > 0`

File: lib/ansible/modules/cloud/amazon/ses_rule_set.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    name = module.params.get('name').lower()
+    state = module.params.get('state').lower()
+    activate = module.params.get('activate')
+    check_mode = module.check_mode
+    changed = False
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='Python module "boto3" is missing, please install it')
+
+    if not HAS_BOTOCORE:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ses_rule_set.py]: just use HAS_BOTO3

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by tomislacker** [comment]
Thank you very much @willthames for the great suggestions, I've since applied those as well as the resolutions for my linting failures (admittedly I copied pretty verbosely from a few ageing Amazon modules).

**Comment by tomislacker** [comment]
ready_for_review

**Comment by tomislacker** [comment]
Anyone have a chance to review these?

File: lib/ansible/modules/cloud/amazon/ses_rule.py (hunk 2 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/ses_rule.py]: Looks like you went a bit far addressing this - now you don't import botocore at all

You need the above code

File: lib/ansible/modules/cloud/amazon/ses_rule.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+                'Actions': actions,
+                'ScanEnabled': scan_enabled,
+            },
+        }
+
+        if not rule_exists(client, ruleset, name):
+            # Rule doesn't exist, add it
+            changed = True
+
+
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ses_rule.py]: Too many blank lines here.

**Comment by tomislacker** [comment]
ready_for_review

**Comment by tomislacker** [comment]
ready_for_review

File: lib/ansible/modules/cloud/amazon/ses_rule_set.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+                active_name = ""
+
+            if name != active_name:
+                changed = True
+
+                if not check_mode:
+                    try:
+                        client.set_active_receipt_rule_set(
+                            RuleSetName=name)
+                    except (botocore.exceptions.ParamValidationError, botocore.exceptions.ClientError) as e:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ses_rule_set.py]: See the [guidelines](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#boto3-2) for better exception handling. You'll need to handle ClientError separately to ParamValidationError as only the former has a response attribute.

So 

```
try:
    result = connection.aws_call()
except ClientError as e:
    module.fail_json(msg=e.message, exception=traceback.format_exc(),
                     **camel_dict_to_snake_dict(e.response))
except ParamValidationError as e:
    module.fail_json(msg=e.message, exception=traceback.format_exc())
```

File: lib/ansible/modules/cloud/amazon/ses_rule.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    return len([r for r in rules if r['Name'].lower() == name.lower()]) > 0
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        name=dict(type='str', required=True),
+        ruleset=dict(type='str', required=True),
+        state=dict(type='str', default='present', choices=['present', 'absent']),
+        after=dict(type='str', required=False),
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ses_rule.py]: `type='str'` and `required=False` are both defaults, and neither is required in the argument spec (this applies anywhere either might occur)

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
I wish I'd spotted the exception handling earlier. I can say that I'll be happy with this once that's fixed - I can't see any other blocking problems.

**Comment by tomislacker** [comment]
@willthames thanks for the additional input, your most recent comments have been resolved.

ready_for_review

**Comment by willthames** [comment]
shipit

thanks @tomislacker

**Comment by tomislacker** [comment]
@willthames is the `needs_revision` tag preventing this from getting merged? Or is there something else I need to do get move things along?

**Comment by willthames** [comment]
@tomislacker looks like the CI hasn't run since you fixed the issues. The easiest way to do that is to close and reopen the PR.

This probably needs more shipits, I think once ansibot notices the CI passing it will notify other people that there's a new module ready for review.

**Pull request closed without merging**